gdxsv/lbs_discord.go

```diff
@@ -0,0 +1,494 @@
+			}
+
+			lobbyFields = append(lobbyFields, &discordEmbedField{
+				Name:  name,
+				Value: v,
+			})
+		}
+	}
+
+	//Following Fields would be all lobbies
+	//Use sortedKeys to fix the ordering
+	for _, i := range sortedKeys(lobby) {
+		l := lobby[i]
+
+		value := addBlankIfRequired(l.RenpoPeers+l.ZeonPeers) + addBlankIfRequired(l.RenpoRoomPeers+l.ZeonRoomPeers) + addBlankIfRequired(l.NoForcePeers)
+
+		for i, v := range splitEmbedFieldValues(value) {
+			name := "⠀"
+			if i == 0 {
+				name = fmt.Sprintf("**%s － %d 人\n%s %s**", l.Name, l.Count, l.RegionName, l.Comment)
+			}
+
+			lobbyFields = append(lobbyFields, &discordEmbedField{
+				Name:  name,
+				Value: v,
+			})
+		}
+
+	}
+	for i, fields := range splitEmbedFields(lobbyFields) {
+		description := ""
+		if i == 0 {
+			description = fmt.Sprintf("🌐 **待機中 %d 人**", lobbyPeerCount)
+		}
+		payload.Embed = append(payload.Embed, &discordEmbed{
+			Description: description,
+			Color:       24041,
+			Fields:      fields,
+		})
+	}
+
+	//
+	//3rd Embed, battle count
+	//
+	var battleFields []*discordEmbedField
+	for _, b := range battle {
+
+		value := addBlankIfRequired(b.RenpoPeers + b.ZeonPeers)
+
+		for i, v := range splitEmbedFieldValues(value) {
+			name := "⠀"
+			if i == 0 {
+				name = b.RegionName
+			}
+
+			battleFields = append(battleFields, &discordEmbedField{
+				Name:  name,
+				Value: v,
+			})
+		}
+	}
+	for i, fields := range splitEmbedFields(battleFields) {
+		description := ""
+		if i == 0 {
+			description = fmt.Sprintf("💥 **戦闘中 %d 人**", battlePeerCount)
+		}
+		payload.Embed = append(payload.Embed, &discordEmbed{
+			Description: description,
+			Color:       13179394,
+			Fields:      fields,
+		})
+	}
+
+	//
+	// Create the json
+	//
+	var jsonData []byte
+	jsonData, err := json.Marshal(payload)
+
+	if err != nil {
+		logger.Error("Failed to create Discord JSON", zap.Error(err))
+		return
+	}
+	jsonString := string(jsonData)
+	logger.Info(jsonString, zap.Int("length", len(jsonString)))
+
+	send(jsonData)
+
+}
+
+//
+// Send to Discord
+//
+func send(jsonData []byte) {
+
+	req, err := http.NewRequest("PATCH", conf.DiscordWebhookURL, bytes.NewBuffer(jsonData))
+	req.Header.Set("Content-Type", "application/json")
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		logger.Error("Failed to publish to Discord", zap.Error(err))
+		return
+	}
+	defer resp.Body.Close()
+
+	logger.Info("Discord Webhook sent", zap.String("Status", resp.Status))
+	if resp.Status == "400 Bad Request" {
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			logger.Error("Failed to read response body", zap.Error(err))
+			return
+		}
+		logger.Error("Failed to create Discord JSON", zap.String("Error:", string(body)))
+	} else if resp.Status == "429 Too Many Requests" {
+
+		resetepochTime := resp.Header.Get("x-ratelimit-reset")
+		sec, _ := strconv.ParseInt(resetepochTime, 10, 64)
+		logger.Info("Rate limit", zap.Int64("x-ratelimit-reset", sec))
```

Comment:
We have not to update as often as caring about rate limits.
Check a flag every second to see if the status is updated and send it if necessary.